File: client/web/src/types/messages.ts
export const MessageType = {
  HELLO: "hello",
  ERROR: "error",
  CHAT_SEND: "chat_send",
  REACT_ADD: "chat_react_add",
  REACT_REMOVE: "chat_react_remove",
  CHANNEL_JOIN: "channel_join",
  CHANNEL_LEAVE: "channel_leave",
  CHANNEL_MEMBERS: "channel_members",
  // Future types go here
} as const;

export type MessageTypeValue = typeof MessageType[keyof typeof MessageType];

export interface BaseMessage {
  type: MessageTypeValue;
  timestamp: string;
  id?: string; // Changed from correlation_id to id to match backend
  payload: any;
}

// User information object (matches backend UserFrom)
export interface UserFrom {
  username: string;
}

// Hello (Client → Server: send token only, no username)
export interface HelloPayloadClientToServer {
  token?: string; // Optional JWT token for authentication
}

export interface HelloMessageClientToServer extends BaseMessage {
  type: typeof MessageType.HELLO;
  payload: HelloPayloadClientToServer;
}

// Hello (Server → Client: returns user info, especially for guests)
export interface HelloPayloadServerToClient {
  token?: string;
  user?: UserFrom; // Server returns the assigned username
}

export interface HelloMessageServerToClient extends BaseMessage {
  type: typeof MessageType.HELLO;
  payload: HelloPayloadServerToClient;
}

// Chat Send (Client → Server)
export interface ChatSendPayloadClientToServer {
  channel_id: number;
  content: string;
}

export interface ChatSendMessageClientToServer extends BaseMessage {
  type: typeof MessageType.CHAT_SEND;
  payload: ChatSendPayloadClientToServer;
}

// Chat Send (Server → Client: broadcasts to channel with sender info)
export interface ChatSendPayloadServerToClient {
  channel_id: number;
  sender?: UserFrom; // User who sent the message
  content: string;
}

export interface ChatSendMessageServerToClient extends BaseMessage {
  type: typeof MessageType.CHAT_SEND;
  payload: ChatSendPayloadServerToClient;
}

// Channel Join (Client → Server)
export interface ChannelJoinPayloadClientToServer {
  channel_id: number;
  user?: UserFrom; // Optional user info
}

export interface ChannelJoinMessageClientToServer extends BaseMessage {
  type: typeof MessageType.CHANNEL_JOIN;
  payload: ChannelJoinPayloadClientToServer;
}

// Channel Join (Server → Client: broadcasts when user joins)
export interface ChannelJoinPayloadServerToClient {
  channel_id: number;
  user?: UserFrom; // User who joined
}

export interface ChannelJoinMessageServerToClient extends BaseMessage {
  type: typeof MessageType.CHANNEL_JOIN;
  payload: ChannelJoinPayloadServerToClient;
}

// Channel Leave (Server → Client only)
export interface ChannelLeavePayload {
  channel_id: number;
  user?: UserFrom; // User who left
}

export interface ChannelLeaveMessage extends BaseMessage {
  type: typeof MessageType.CHANNEL_LEAVE;
  payload: ChannelLeavePayload;
}

// Error (server sends this)
export interface ErrorPayload {
  detail: string;
}

export interface ErrorMessage extends BaseMessage {
  type: typeof MessageType.ERROR;
  payload: ErrorPayload;
}

// Reactions (Client → Server)
export interface ReactPayloadClientToServer {
  emote: string; // Emoji string
  message_id: string; // UUID of the message being reacted to
  channel_id: number;
}

export interface ReactAddMessageClientToServer extends BaseMessage {
  type: typeof MessageType.REACT_ADD;
  payload: ReactPayloadClientToServer;
}

export interface ReactRemoveMessageClientToServer extends BaseMessage {
  type: typeof MessageType.REACT_REMOVE;
  payload: ReactPayloadClientToServer;
}

// Reactions (Server → Client: broadcasts to channel)
export interface ReactPayloadServerToClient {
  emote: string;
  message_id: string;
  channel_id: number;
}

export interface ReactAddMessageServerToClient extends BaseMessage {
  type: typeof MessageType.REACT_ADD;
  payload: ReactPayloadServerToClient;
}

export interface ReactRemoveMessageServerToClient extends BaseMessage {
  type: typeof MessageType.REACT_REMOVE;
  payload: ReactPayloadServerToClient;
}

// Channel Members (Server → Client: sent when member list changes)
export interface ChannelMembersPayload {
  channel_id: number;
  members: UserFrom[];
}

export interface ChannelMembersMessage extends BaseMessage {
  type: typeof MessageType.CHANNEL_MEMBERS;
  payload: ChannelMembersPayload;
}

// Union type for messages received from server
export type Message =
  | HelloMessageServerToClient
  | ChatSendMessageServerToClient
  | ChannelJoinMessageServerToClient
  | ChannelLeaveMessage
  | ErrorMessage
  | ReactAddMessageServerToClient
  | ReactRemoveMessageServerToClient
  | ChannelMembersMessage;

File: server/src/chat_server/protocol/messages.py
from typing import Literal

from pydantic import UUID4, BaseModel, ConfigDict

from chat_server.protocol.basemessage import BaseMessage
from chat_server.protocol.enums import MessageType
from chat_server.protocol.registry import register_message


class UserFrom(BaseModel):
    """
    User information to be transported.
    """

    username: str
    model_config = ConfigDict(from_attributes=True)


# Error
class ErrorMessagePayload(BaseModel):
    detail: str


@register_message(MessageType.ERROR)
class ErrorMessage(BaseMessage):
    type: Literal[MessageType.ERROR] = MessageType.ERROR
    payload: ErrorMessagePayload


# Hello
class HelloPayload(BaseModel):
    model_config = {"extra": "forbid"}
    token: str | None = None
    user: UserFrom | None = None


@register_message(MessageType.HELLO)
class Hello(BaseMessage):
    type: Literal[MessageType.HELLO] = MessageType.HELLO
    payload: HelloPayload


# Channel Join
class ChannelJoinPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    user: UserFrom | None = None  # Server-only


@register_message(MessageType.CHANNEL_JOIN)
class ChannelJoin(BaseMessage):
    type: Literal[MessageType.CHANNEL_JOIN] = MessageType.CHANNEL_JOIN
    payload: ChannelJoinPayload


# Channel Leave
class ChannelLeavePayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    user: UserFrom | None = None  # Server-only


@register_message(MessageType.CHANNEL_LEAVE)
class ChannelLeave(BaseMessage):
    type: Literal[MessageType.CHANNEL_LEAVE] = MessageType.CHANNEL_LEAVE
    payload: ChannelLeavePayload


# Chat Send
class ChatSendPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    sender: UserFrom | None = None
    content: str


@register_message(MessageType.CHAT_SEND)
class ChatSend(BaseMessage):
    type: Literal[MessageType.CHAT_SEND] = MessageType.CHAT_SEND
    payload: ChatSendPayload


# Reacts
class ReactPayload(BaseModel):
    model_config = {"extra": "forbid"}
    # TODO: Limit the possible values for an emote
    emote: str
    message_id: UUID4
    channel_id: int


@register_message(MessageType.REACT_ADD)
class ReactAdd(BaseMessage):
    type: Literal[MessageType.REACT_ADD] = MessageType.REACT_ADD
    payload: ReactPayload


@register_message(MessageType.REACT_REMOVE)
class ReactRemove(BaseMessage):
    type: Literal[MessageType.REACT_REMOVE] = MessageType.REACT_REMOVE
    payload: ReactPayload


# Channel Members
class ChannelMembersPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    members: list[UserFrom]


@register_message(MessageType.CHANNEL_MEMBERS)
class ChannelMembers(BaseMessage):
    type: Literal[MessageType.CHANNEL_MEMBERS] = MessageType.CHANNEL_MEMBERS
    payload: ChannelMembersPayload


# Typing Start
class TypingStartPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    user: UserFrom | None = None  # Ignored by the server if the client send this.


@register_message(MessageType.TYPING_START)
class TypingStart(BaseMessage):
    type: Literal[MessageType.TYPING_START] = MessageType.TYPING_START
    payload: TypingStartPayload


#################
# Chat Commands #
#################
class KickCommandPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    target: str
    reason: str = ""


@register_message(MessageType.CHAT_KICK)
class KickCommand(BaseMessage):
    type: Literal[MessageType.CHAT_KICK] = MessageType.CHAT_KICK
    payload: KickCommandPayload


class MuteCommandPayload(BaseModel):
    model_config = {"extra": "forbid"}
    channel_id: int
    target: str
    duration: int | None = None
    reason: str = ""


@register_message(MessageType.CHAT_MUTE)
class MuteCommand(BaseMessage):
    type: Literal[MessageType.CHAT_MUTE] = MessageType.CHAT_MUTE
    payload: MuteCommandPayload
